Support custom xDS channel creds

yousukseung opened the pull request:
This patch introduces a factory to allow supporting custom xDS channel
creds. Three types currently supported (fake, insecure, google_default)
are registered by default for backward-compatibility.




<!--

Your pull request will be routed to the following person by default for triaging.
If you know who should review your pull request, please remove the mentioning below.

-->

@markdroth, @ctiller

markdroth commented on src/core/ext/xds/xds_bootstrap.cc:
```diff
@@ -98,39 +100,118 @@ grpc_error_handle ParseChannelCredsArray(const Json::Array& json,
+}
+
+void XdsChannelCredsRegistry::Init() {
+  if (g_state == nullptr) g_state = new XdsChannelCredsRegistryState();
```
I think we should just unconditionally set it here, without a check to see that it's null.

markdroth commented on src/core/ext/xds/xds_bootstrap.cc:
```diff
@@ -98,39 +100,118 @@ grpc_error_handle ParseChannelCredsArray(const Json::Array& json,
+
+void XdsChannelCredsRegistry::Shutdown() {
+  delete g_state;
+  g_state = nullptr;
```
Resetting to null should not be needed.

markdroth reviewed (changes requested):
Thanks for doing this!

It looks like you've patterned this after some of our older registries.  I think the one you should use as a model here is the one in xds_http_filters.{h,cc}, since that's both the most recent and the most similar to what you're doing here.

Please let me know if you have any questions.  Thanks!

yousukseung commented:
> 

Thanks, I chat with Craig and he was OK to proceed with the old style for now and revisit.
@ctiller, are you still OK with this change?

yousukseung commented:
> I suspect that we're talking about two different things here.
> 
> We have previously discussed the work that Craig is doing on the new CoreConfiguration mechanism, and the fact that we will eventually need to move this registry over to that new mechanism. We agreed that it was fine to create this as an old-style registry for now, and I am not attempting to revisit that decision.
> 
> What I am saying is that some of our older code predated our ability to use things like `std::map<>` inside of C-core, so it used hacks like simulating a map on top of an `InlinedVector<>`. That's not necessary here, so please pattern this code after the newer example I pointed you at. It's still an old-style registry (not using the new CoreConfiguration mechanism), but it can be implemented in a more modern way.

Sorry that I was not clear, I also meant the former. I am already working on addressing your comments will update soon.

yousukseung commented on src/core/ext/xds/xds_bootstrap.h:
```diff
@@ -39,12 +39,25 @@ namespace grpc_core {
+      const Json& config) const = 0;
+  virtual ~XdsChannelCredsFactory() {}
+  virtual bool IsValidConfig(const Json& config) const = 0;
+  virtual const char* creds_type() const = 0;
```
Done.

yousukseung commented on src/core/ext/xds/xds_bootstrap.h:
```diff
@@ -39,12 +39,25 @@ namespace grpc_core {
 class XdsClient;
 
+class XdsChannelCredsFactory {
+ public:
```
Done.

yousukseung commented on src/core/ext/xds/xds_bootstrap.cc:
```diff
@@ -98,39 +100,118 @@ grpc_error_handle ParseChannelCredsArray(const Json::Array& json,
                                        &error_list);
 }
 
+class XdsChannelCredsRegistryState {
```
Done, but went with flat_hash_map since we don't care the order, let me know this does not work.

yousukseung commented on src/core/ext/xds/xds_bootstrap.cc:
```diff
@@ -98,39 +100,118 @@ grpc_error_handle ParseChannelCredsArray(const Json::Array& json,
 //
 
+XdsChannelCredsRegistryState::XdsChannelCredsRegistryState() {
+  RegisterXdsChannelCredsFactory(
```
Done.

markdroth commented on src/core/ext/xds/xds_channel_creds.cc:
```diff
@@ -0,0 +1,109 @@
+namespace {
+
+using ChannelCredsMap =
+    absl::flat_hash_map<absl::string_view,
```
`std::map<>` should be fine here.  There are only ever going to be a small number of entries, so the overhead of hashing is probably overkill.

yousukseung replied:
Sure, done.

markdroth commented on src/core/ext/xds/xds_channel_creds.cc:
```diff
@@ -0,0 +1,109 @@
+        grpc_google_default_credentials_create(nullptr));
+  }
+  bool IsValidConfig(const Json& /*config*/) const override { return true; }
+  absl::string_view creds_type() const override { return "google_default"; }
```
Please define this method first in each of these classes.

yousukseung replied:
Sorry that I missed this, fixed.

markdroth reviewed (commented):
This looks really good!  Just a couple of minor nits remaining.

Please let me know if you have any questions.  Thanks!

markdroth reviewed (approved):
This looks good!

For future reference, please don't force-push to a PR, because that makes it harder for the reviewers to see what's changed since their last review pass.  Thanks!

yousukseung commented:
> 

Thanks, should I create a separate commit or can just amend and push instead of force-push?

markdroth commented:
Just add new commits whenever you make a change.  The commits will all be squashed when the PR is merged.

yousukseung commented:
> Just add new commits whenever you make a change. The commits will all be squashed when the PR is merged.

Got it, thanks.
